Core/include/Acts/TrackFinding/CombinatorialKalmanFilter.hpp

```diff
@@ -529,105 +571,178 @@ class CombinatorialKalmanFilter {
           return boundStateRes.error();
         }
         auto boundState = *boundStateRes;
-        const auto& boundParams = std::get<BoundTrackParameters>(boundState);
+        const auto& [boundParams, jacobian, pathLength] = boundState;
+
+        // Retrieve the previous tip and its state
+        // The states created on this surface will have the common previous tip
+        size_t prevTip = SIZE_MAX;
+        TipState prevTipState;
+        if (not result.activeTips.empty()) {
+          prevTip = result.activeTips.back().first;
+          prevTipState = result.activeTips.back().second;
+          // New state is to be added. Remove the last tip from active tips
+          result.activeTips.erase(result.activeTips.end() - 1);
+        }
 
         // Get all source links on the surface
         auto [lower_it, upper_it] =
             m_sourcelinkAccessor.range(surface->geometryId());
+
+        result.trackStateCandidates.clear();
+        result.trackStateCandidates.reserve(std::distance(lower_it, upper_it));
+
+        result.stateBuffer.clear();
+
+        using PM = TrackStatePropMask;
+
         // Calibrate all the source links on the surface since the selection has
         // to be done based on calibrated measurement
-        std::vector<BoundVariantMeasurement> measurements;
-        measurements.reserve(nSourcelinks);
         for (auto it = lower_it; it != upper_it; ++it) {
-          measurements.emplace_back(
-              m_calibrator(m_sourcelinkAccessor.at(it), boundParams));
+          // get the source link
+          const auto& sourceLink = m_sourcelinkAccessor.at(it);
```

Comment:
As this method is getting large, do you think we could put the creation of the track state candidates into a separate method?

Reply:
Good idea. Factored out two functions `createSourceLinkTrackStates` and `processSelectedTrackStates`.